### ISSUE TITLE
Possible false positive -Wempty-body when indented semicolon after statement

### ISSUE THREAD

**Issue by arurke** [llvm/llvm-project#56452]
Clang-14 with `-Wall` will raise a `-Wempty-body` warning if there is an indented `;` after a if- or while-statement:
```c
while(1);
    ;
```
```console
<source>:6:13: warning: while loop has empty body [-Wempty-body]
    while(1);
            ^
<source>:6:13: note: put the semicolon on a separate line to silence this warning
```
This becomes annoying in use-cases where defines are used to exclude functions:

```c
/* foo.h */
#if OPT_FUNC_NEEDED
void optional_function();
#else
#define optional_function()
#endif

/* bar.c */
#include "foo.h"
void raising_warning(void) {

    while(1);
    optional_function(); // Raises -Wempty-body when !OPT_FUNC_NEEDED

    // The call above is equivalent to the following
    //                 ;
}
```
MWE: https://godbolt.org/z/oG7nqPa8d

Is this a false positive, or intended behavior?

**Comment by inclyc** [comment]
https://godbolt.org/z/h3q1KhM3K

```C++
void raising_warning(void) {
    while(1);
    ; // no warning

    while(1);
        ; // -Wempty-body
    
    while(1)
        ; // no warning
    
    while(1)
    ; // no warning
}
```

Note that the first test case and the second only differ in indentation.

**Comment by inclyc** [comment]
# What's the logic behind this diagnosis?
Copied from `SemaChecking.cpp Sema::DiagnoseEmtpyLoopBody`.

`for(...);` and `while(...);` are popular idioms, so in order to keep
noise level low, emit diagnostics only if for/while is followed by a
`CompoundStmt`, e.g.:
```C++
for (int i = 0; i < n; i++);
{
  a(i);
}
```
or if for/while is followed by a statement with more indentation 
than for/while itself:
```C++
for (int i = 0; i < n; i++);
    a(i);
```

# Is this an intended behavior?

> Clang-14 with `-Wall` will raise a `-Wempty-body` warning if there is an indented `;` after a if- or while-statement:
> 
> ```c
> while(1);
>     ;
> ```

In this case this might be an intended behavior

**Comment by inclyc** [comment]
> ```c
> /* foo.h */
> #if OPT_FUNC_NEEDED
> void optional_function();
> #else
> #define optional_function()
> #endif
> 
> /* bar.c */
> #include "foo.h"
> void raising_warning(void) {
> 
>     while(1);
>     optional_function(); // Raises -Wempty-body when !OPT_FUNC_NEEDED
> 
>     // The call above is equivalent to the following
>     //                 ;
> }
> ```
> 
> MWE: https://godbolt.org/z/oG7nqPa8d
> 
> Is this a false positive, or intended behavior?

Note that:

>     // The call above is equivalent to the following
>     //                 ;

That's absolutely correct, the statement is indented differently than the `while` statement above, because the information about the `optional_function` has been lost in the C preprocessor.
